File: example/java/jimmer-sql/src/main/java/org/babyfish/jimmer/sql/example/cache/MaxwellListener.java
package org.babyfish.jimmer.sql.example.cache;

import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.datatype.jsr310.JavaTimeModule;
import org.babyfish.jimmer.sql.JSqlClient;
import org.babyfish.jimmer.sql.cache.Caches;
import org.springframework.boot.autoconfigure.condition.ConditionalOnProperty;
import org.springframework.kafka.annotation.KafkaListener;
import org.springframework.kafka.support.Acknowledgment;
import org.springframework.stereotype.Component;

@ConditionalOnProperty("spring.redis.host")
@Component
public class MaxwellListener {

    private static final ObjectMapper MAPPER = new ObjectMapper()
            .registerModule(new JavaTimeModule());

    private final Caches caches;

    public MaxwellListener(JSqlClient sqlClient) {
        this.caches = sqlClient.getCaches();
    }

    @KafkaListener(topics = "maxwell")
    public void onHandle(
            String json,
            Acknowledgment acknowledgment
    ) throws JsonProcessingException {
        JsonNode node = MAPPER.readTree(json);
        String tableName = node.get("table").asText();
        if (caches.isAffectedBy(tableName)) {
            String type = node.get("type").asText();
            JsonNode data = node.get("data");
            switch (type) {
                case "insert":
                    caches.invalidateByBinLog(tableName, null, data);
                    break;
                case "update":
                    caches.invalidateByBinLog(tableName, node.get("old"), data);
                    break;
                case "delete":
                    caches.invalidateByBinLog(tableName, data, null);
                    break;
            }
        }
        acknowledgment.acknowledge();
    }
}

File: example/kotlin/jimmer-sql-kt/src/main/kotlin/org/babyfish/jimmer/example/kt/sql/cache/MaxwellListener.kt
package org.babyfish.jimmer.example.kt.sql.cache

import com.fasterxml.jackson.databind.JsonNode
import com.fasterxml.jackson.databind.ObjectMapper
import com.fasterxml.jackson.databind.node.ObjectNode
import org.babyfish.jimmer.sql.kt.KCaches
import org.babyfish.jimmer.sql.kt.KSqlClient
import org.springframework.boot.autoconfigure.condition.ConditionalOnProperty
import org.springframework.kafka.annotation.KafkaListener
import org.springframework.kafka.support.Acknowledgment
import org.springframework.stereotype.Component


@ConditionalOnProperty("spring.redis.host")
@Component
class MaxwellListener(sqlClient: KSqlClient) {

    private val caches: KCaches = sqlClient.caches

    @KafkaListener(topics = ["maxwell"])
    fun onHandle(
        json: String,
        acknowledgment: Acknowledgment
    ) {
        val node = MAPPER.readTree(json)
        val tableName = node["table"].asText()
        if (caches.isAffectedBy(tableName)) {
            val type = node["type"].asText()
            val data = node["data"]
            when (type) {
                "insert" ->
                    caches.invalidateByBinLog(tableName, null, data)
                "update" -> {
                    val old = node["old"] as ObjectNode
                    old.set<JsonNode>("id", data["id"])
                    caches.invalidateByBinLog(tableName, old, data)
                }
                "delete" ->
                    caches.invalidateByBinLog(tableName, data, null)
            }
        }
        acknowledgment.acknowledge()
    }

    companion object {
        private val MAPPER = ObjectMapper()
    }
}